062-UseKRaft-feature-gate-promotion.md

```diff
@@ -0,0 +1,94 @@
+* Support for migration of ZooKeeper-based clusters to KRaft
+  * Tracked in [strimzi/strimzi-kafka-operator#9433](https://github.com/strimzi/strimzi-kafka-operator/issues/9433)
+  * This is currently work in progress and waiting for Apache Kafka 3.7.0 release with several bug-fixes and implementation of [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum+and+add+Controller+Registration)
+  * Support for the migration is critical for existing users and has to be done before support for ZooKeeper-based cluster is dropped.
+    But missing support for migration does not affect any users who would want to deploy and run new clusters in KRaft mode.
+  * This is currently expected to be implemented in Strimzi 0.40 or 0.41.
+* Scaling of KRaft controller nodes
+  * Tracked in [strimzi/strimzi-kafka-operator#9429](https://github.com/strimzi/strimzi-kafka-operator/issues/9429)
+  * Currently not supported are:
+    * Scale-down of dedicated or mixed controller nodes that breaches the controller quorum
+    * Scale-up of dedicated controller nodes (the scale-up eventually succeeds, but the broker availability is not maintained)
+  * This is currently blocked by the support for [KIP-853](https://cwiki.apache.org/confluence/display/KAFKA/KIP-853%3A+KRaft+Controller+Membership+Changes).
+    This KIP might not be implemented earlier than in Apache Kafka 4.0.
+    Support for scaling controllers is currently not seen as blocker for moving the `UseKRaft` feature gate to _GA_.
+* JBOD support
+  * Tracked in [strimzi/strimzi-kafka-operator#9437](https://github.com/strimzi/strimzi-kafka-operator/issues/9437)
+  * Expected to be implemented in Apache Kafka 3.7.0.
+    The Strimzi implementation fill follow the release of Kafka 3.7.0 in Strimzi 0.40 or 0.41.
+
+## Proposed timeline
+
+This proposal proposes the following timeline for the graduation of the `UseKRaft` feature gate:
+* Move to _beta_ phase and be enabled by default in Strimzi 0.40.0
+* Move to _GA_ phase and be permanently enabled in Strimzi 0.42.0
+
+It is worth noting that:
+* In addition to the `UseKRaft` feature gate, the KRaft clusters are enabled / disabled using the `strimzi.io/kraft` annotation.
+  As a result, moving the `UseKRaft`` feature gate to _beta_ or _GA_ does not mean that all new Kafka clusters have to use KRaft or that this has any impact on existing ZooKeeper based clusters.
+  KRaft will be applied only to the Kafka clusters with the right annotation.
+* Moving the `UseKRaft` feature to beta or GA does not mean we will drop support for ZooKeeper-based clusters.
+  While moving the `UseKRaft` feature gate to _GA_ defines the earliest moment when support for ZooKeeper based clusters can be dropped, it is currently not expected to happen in Strimzi 0.42 and this proposal does not establish any such plan.
```

Comment:
This is in contradiction with the previous phrase, as you say it's not expected, which means it is unlikely but can happen.

I think we want to have both KRaft and ZK supported as GA in at least one release. There are a lot of users that would not consider the migration until KRaft support is declared GA in Strimzi. If we want to support that, the earliest moment to drop ZK (and migration) would be 0.43. This is also not expected to happen IMO, as we will probably have multiple releases with both ZK and KRaft modes supported as GA.

Reply:
Where is it a contradiction? It says that `UseKRaft` being GA is only one of the prerequisites of ZooKeeper support being removed, that ZooKeeper support removal is currently not expected to happen in 0.42 and that no timeline for it is established by this PR. None of that contradicts each other.

Reply:
not expected != won't happen, I think we should be clear about that

Reply:
That is correct and as the second part of the sentence explains, that is not for this proposal to guarantee or define one way or another in any way. For that to happen, there would be another proposal.

Reply:
> not for this proposal to guarantee or define one way or another

exactly, but this is not what line 42 says

Reply:
How is it not in line? They are fully in sync from my perspective.

Reply:
My problem is with the words "not expected", which IMO do not exclude the possibility of another proposal dropping support for ZK when KRaft becomes GA. 

Can't we simply say that ZK support will be dropped at some point in time *after* KRaft will be declared GA in Strimzi? I think that would be much clearer.

Reply:
With "right after" is ok for me. Thanks.